File: src/abstractions/Bakabase.Abstractions/Models/Domain/Resource.cs
using Bakabase.InsideWorld.Models.Constants;
using System.Collections.Frozen;
using Bakabase.Abstractions.Extensions;
using Bakabase.Abstractions.Models.Domain.Constants;

namespace Bakabase.Abstractions.Models.Domain;

public record Resource
{
    public int Id { get; set; }

    [Obsolete]
    public int MediaLibraryId { get; set; }

    public ResourceStatus Status { get; set; } = ResourceStatus.Active;

    /// <summary>
    /// Source links for this resource. A resource can be discovered by multiple sources.
    /// </summary>
    public List<ResourceSourceLink>? SourceLinks { get; set; }

    public string? FileName => string.IsNullOrEmpty(Path) ? null : System.IO.Path.GetFileName(Path);

    public string? Directory => string.IsNullOrEmpty(Path) ? null : System.IO.Path.GetDirectoryName(Path).StandardizePath()!;

    public string? Path { get; set; }

    private string? _displayName;

    public string? DisplayName
    {
        get => _displayName ?? FileName;
        set => _displayName = value;
    }

    public int? ParentId { get; set; }
    public bool HasChildren => Tags.Contains(ResourceTag.IsParent);
    public bool IsFile { get; set; }

    public DateTime CreatedAt { get; set; } = DateTime.Now;
    public DateTime UpdatedAt { get; set; } = DateTime.Now;
    public DateTime FileCreatedAt { get; set; }
    public DateTime FileModifiedAt { get; set; }
    public List<string>? CoverPaths { get; set; }
    public IReadOnlySet<ResourceTag> Tags { get; set; } = FrozenSet<ResourceTag>.Empty;
    public Resource? Parent { get; set; }
    /// <summary>
    /// ResourcePropertyType - PropertyId - Property
    /// </summary>
    public Dictionary<int, Dictionary<int, Property>>? Properties { get; set; }

    public bool Pinned => Tags.Contains(ResourceTag.Pinned);
    public DateTime? PlayedAt { get; set; }

    /// <summary>
    /// Final resolved cover paths for this resource, populated by service layer using priority:
    /// 1. User-set covers (ReservedProperty.Cover, scope=Manual)
    /// 2. External source local covers (SourceLink.LocalCoverPaths)
    /// 3. Enhancer covers (scope=XxxEnhancer)
    /// 4. FileSystem auto-discovered covers (from cache)
    /// </summary>
    public List<string>? Covers { get; set; }

    /// <summary>
    /// Final resolved playable items for this resource, aggregated from all sources.
    /// </summary>
    public List<PlayableItem>? PlayableItems { get; set; }

    /// <summary>
    /// Whether there are more FileSystem playable items beyond what's shown (due to trimming limits).
    /// </summary>
    public bool HasMoreFileSystemPlayableItems { get; set; }

    /// <summary>
    /// Whether covers have been resolved and are ready.
    /// </summary>
    public bool CoversReady { get; set; }

    /// <summary>
    /// Whether playable items have been resolved and are ready.
    /// </summary>
    public bool PlayableItemsReady { get; set; }

    public ResourceCache? Cache { get; set; }

    public record Property(
        string? Name,
        PropertyType Type,
        List<Property.PropertyValue>? Values,
        bool Visible = false,
        int Order = 0)
    {
        public string? Name { get; set; } = Name;
        public List<PropertyValue>? Values { get; set; } = Values;
        public PropertyType Type { get; set; } = Type;
        public StandardValueType DbValueType => PropertyTypeValueTypes.GetDbValueType(Type);
        public StandardValueType BizValueType => PropertyTypeValueTypes.GetBizValueType(Type);
        public bool Visible { get; set; } = Visible;
        public int Order { get; set; } = Order;

        public record PropertyValue(
            int Scope,
            object? Value,
            object? BizValue,
            object? AliasAppliedBizValue)
        {
            public int Scope { get; set; } = Scope;
            public object? Value { get; set; } = Value;
            public object? BizValue { get; set; } = BizValue ?? Value;
            public object? AliasAppliedBizValue { get; set; } = AliasAppliedBizValue ?? BizValue ?? Value;

            public bool IsManuallySet => Scope == (int)PropertyValueScope.Manual;
        }
    }

    [Obsolete]
    public string? MediaLibraryName { get; set; }
    [Obsolete]
    public string? MediaLibraryColor { get; set; }

    public List<MediaLibraryInfo>? MediaLibraries { get; set; }

    public record MediaLibraryInfo(int Id, string Name, string? Color);
}

File: src/web/src/core/models/Resource.ts
import type {
  DataOrigin,
  DataStatus,
  PropertyType,
  PropertyValueScope,
  ResourceDataType,
  ResourceSource,
  ResourceStatus,
  ResourceTag,
  StandardValueType,
} from "@/sdk/constants";
import type { PropertyPool } from "@/sdk/constants";

type Value = {
  scope: PropertyValueScope;
  value?: any;
  bizValue?: any;
  aliasAppliedBizValue?: any;
};

/**
 * Whether a scope's resolved value actually holds content. Plain truthiness is wrong here: an
 * empty array is truthy but empty, while 0 and false are falsy but valid. Used to skip empty
 * scopes when resolving a property's multi-scope values down to one displayed value.
 */
export const isNonEmptyValue = (value: any): boolean => {
  if (value == null) return false;
  if (Array.isArray(value)) return value.length > 0;
  if (typeof value === "string") return value.trim().length > 0;
  return true;
};

export type Property = {
  name?: string;
  type: PropertyType;
  dbValueType: StandardValueType;
  bizValueType: StandardValueType;
  values?: Value[];
  visible?: boolean;
  order: number;
};

export type PlayableItem = {
  origin: DataOrigin;
  key: string;
  displayName?: string;
};

export type ResourceDataState = {
  resourceId: number;
  dataType: ResourceDataType;
  origin: DataOrigin;
  status: DataStatus;
};

export type PropertyValueScopePriority = {
  scope: PropertyValueScope;
  /** When this scope is empty, whether to continue with the next scope or stop and render blank. Ignored on the last entry. */
  fallbackOnEmpty: boolean;
};

export type PropertyValueScopePreference = {
  resourceId: number;
  propertyPool: PropertyPool;
  propertyId: number;
  /** Ordered scope priorities with per-scope fallback flag; null = no override (falls through to profile/global) */
  priorities?: PropertyValueScopePriority[];
};

export type ResourceSourceLink = {
  id: number;
  resourceId: number;
  source: ResourceSource;
  sourceKey: string;
  coverUrls?: string[];
  localCoverPaths?: string[];
};

export type Resource = {
  id: number;
  mediaLibraryId: number;
  categoryId: number;
  status: ResourceStatus;
  sourceLinks?: ResourceSourceLink[];
  fileName?: string;
  directory?: string;
  displayName?: string;
  path?: string;
  parentId?: number;
  hasChildren: boolean;
  isFile: boolean;
  createdAt: string;
  updatedAt: string;
  fileCreatedAt: string;
  fileModifiedAt: string;
  parent?: Resource;
  properties?: { [key in PropertyPool]?: Record<number, Property> };
  /** @deprecated */
  mediaLibraryName?: string;
  /** @deprecated */
  mediaLibraryColor?: string;
  mediaLibraries?: { id: number; name: string; color?: string }[];
  /** @deprecated */
  category?: { id: number; name: string };
  pinned: boolean;
  tags: ResourceTag[];
  playedAt?: string;
  /** Aggregated health score; null when no profile has scored this resource. */
  healthScore?: number;
  dataStates?: ResourceDataState[];

  /** Final resolved cover paths, populated by backend using priority-based selection */
  covers?: string[];
  /** Final resolved playable items from all sources */
  playableItems?: PlayableItem[];

  /** Per-(propertyPool, propertyId) scope priority overrides for this resource */
  scopePreferences?: PropertyValueScopePreference[];
};
